### ISSUE TITLE
[feature request] curly brace support

### ISSUE THREAD

**Issue by avilleret** [pure-data/pure-data#505]
hi, 

I saw that backslash are now supported which is great, thanks @millerpuckette 
would it be possible to also support opening and closing curly brace ?
those are really useful in some syntax context like regexp

for example, with libossia, man can control several parameters at once with [ø.remote foo.{1,2,4,15} ]
which is quite convenient.

for now we have a workaround but curly braces are so nice :-)

**Comment by porres** [comment]
It seems at least a prerequisite for this has already been merged, see https://github.com/pure-data/pure-data/pull/647

I was about to request the same feature and saw this. Anyway, I wonder what else is on the way of implementing this, maybe I can try and help doing it...

**Comment by porres** [comment]
@umlaeute what's in the way to support this these days?

**Comment by umlaeute** [comment]
1. technically i guess most hurdles are taken (but then, i haven't really checked). i think the current state is:
 - [x] allow display of curlies in the GUI (core -> gui communication) - this has already been solved
 - [ ] allow input of curlies (gui -> communication) - this is simple to solve it just requires us to drop the filter
 - [ ] escaping of curlies when (re)storing files (they shouldn't be escaped at all); this might already be solved (i haven't checked)

2. however *practically* i think we should **not** lightly add "support for curly braces".
    the fact that curlies (that is: a pair of matching characters) have been forbidden until now is a unique opportunity, as it could allow for a simple and elegant way to write **structured data** in Pd patches.
currently our only data structure we can express within a message are lists of atoms. but with curlies we could introduce  *list of lists* without breaking backward compatibility:

   ```
   [list foo {1 2 3} bar(
   |
   [$1(
   |
   |list 1 2 3)
   ```

   I think this is much more interesting that allowing for regexp-groups such as used by OSC or ossia.

   note that this is not mutually exclusive; we can have both single curlies denoting some special syntax (like the lists-of-lists above) **and** curlies within symbols to not carry any specific meaning (so that a symbol like `/{foo,bar}` can be used  as an OSC-path with alternatives, handled by the OSC-receiver).

**Comment by porres** [comment]
> we can have both single curlies denoting some special syntax (like the lists-of-lists above) and curlies within symbols to not carry any specific meaning

That's what I was thinking. So, the thing is we'd need to use/implement single curlies first as this special syntax before also allowing other stuff?

**Comment by umlaeute** [comment]
yes, i think so.
and before that, we need to agree on the actual use of single curlies (which is probably the hardest part)

**Comment by Spacechild1** [comment]
Copy pasting my reply on pd-list:

> the fact that curlies (that is: a pair of matching characters) have been forbidden until now is a unique opportunity. 
Well, they have not been forbidden. You can not *type* them (yet), but they can appear in text files or in manually assembled symbols. In fact, they are properly displayed with [print] or list/symbol atoms, which suggests that they are already properly escaped when sent to the GUI. One could argue that someone who only uses libpd might not even be aware of this issue.

Curly braces have never been a part of special Pd symbols (semicolon, colon, dollar), so I'm sceptical that we can just give them some meaning. It would break lots of existing code that uses symbols with curly braces. For example, it would completely break the "purest_json" library.

However, we could use *escaped* curly braces. The escape sequences "\{" and "\}" are not defined, so we might use those for nested lists. I'm not saying it's pretty, though.

**Comment by umlaeute** [comment]
> For example, it would completely break the "purest_json" library.

would it? how so?

my proposal distinguished between curlies in symbols and standalone curlies (similar to how Pd distinguishes between commas in symbols and standalone commas).

curlies in symbols would just be treated literally, so a symbol `{"id":42}` would still be parsable by `[json-decode]` (and `[json-encode]` would still be allowed to create such symbols).
otoh `[10 { 1 2 3 } 20(` could contain a 3 element list (with the 2nd element being a list itself), but creating the symbols `{` resp `}` (by arcane magic like `[makefilename %c]`) and feeding them into `[10 $1 1 2 3 $2 20(` would create a flat 7 element list (just like creating two `,`-symbols and feeding them into the $arg expansion would create a 7 element list rather than the 3 messages of `[10 , 1 2 3 , 20(`.

**Comment by Spacechild1** [comment]
> curlies in symbols would just be treated literally, so a symbol {"id":42} would still be parsable by [json-decode] (and [json-encode] would still be allowed to create such symbols).

you're right that purest_json incidentally stores the whole JSON string in a single symbol, but there are other libraries/patches that use symbols with a single curly brace. Personally, I have abstractions that create Lua code, such as `{ foo = 123 \, bar = 567 }`. These are valid Pd messages. 

Again, curly braces have never been a reserved character in the Pd *core*, but we could safely use *escaped* curly braces.

**Comment by Spacechild1** [comment]
> (similar to how Pd distinguishes between commas in symbols and standalone commas).

That's not true., you cannot have an *unescaped* comma (or semicolon) in a symbol. `[symbol foo,bar(` gives you two messages (`symbol foo` and `bar`).  `[symbol foo,bar]` creates a symbol `foo`.

**Comment by Ant1r** [comment]
> use symbols with a single curly brace

Even if the curly brace was chosen to create nested lists, you would still be able to create such symbol by escaping the brace, like with the comma or space.
If you want to create the list: `{ foo = 123 \, bar = 567 }`
you would have to type the message: `[list \{ foo = 123 \, bar = 567 \}(`.

**Comment by reduzent** [comment]
One could argue that the escaping is only necessary for a symbol atom containing nothing else but a curly brace character. Symbols containing curly braces and other characters wouldn't interfere with IOhannes' proposal. 

* `{`, `}` -> list delimiters
* `\{` -> symbolic atom encoding `{`-symbol
* `{sdajflk` -> symbolic atom `{sdajflk`

**Comment by reduzent** [comment]
> > (similar to how Pd distinguishes between commas in symbols and standalone commas).
> 
> That's not true, you cannot have an _unescaped_ comma (or semicolon) in a symbol _in a Pd message_. `[symbol foo,bar(` gives you two messages (`symbol foo` and `bar`). `[symbol foo,bar]` creates a symbol `foo`.

You can type `foo,bar` into a symbol box and you get a `foo,bar` symbol. However, it is printed in its FUDI encoded form `symbol foo\,bar`. Nevertheless, the presentation / decoded form is `foo,bar'. When writing message boxes or object boxes, you obviously need to write symbols in their FUDI encoded form.

**Comment by reduzent** [comment]
> Again, curly braces have never been a reserved character in the Pd _core_, but we could safely use _escaped_ curly braces.

I find that a really awkward proposal and goes contrary to well established principles. Usually, escaping is used to _escape_ the special meaning of a character, not vice versa. As long as FUDI-specific special characters are escaped during FUDI-encoding, there should be no issue, even with existing externals.

**Comment by Spacechild1** [comment]
> If you want to create the list: { foo = 123 \, bar = 567 }

The point is, this list is already a valid Pd message! `{` and `}` do not have any special meaning in Pd. (They only need to be escaped when sent back to the Tcl/Tk GUI). Changing the meaning of these characters would break existing patches.

> I find that a really awkward proposal and goes contrary to well established principles. Usually, escaping is used to escape the special meaning of a character, not vice versa. 

I agree that it's not pretty. I conceded that in my first reply. However, it is the *only* way to do this without breaking existing patches.

But @umlaeute's proposal would go against well established principles as well. No FUDI-specific character in a Pd messages cares whether it is surrounded by whitespace. `foo bar,baz 0` is the same as `foo bar , baz 0`.

> As long as FUDI-specific special characters are escaped during FUDI-encoding, there should be no issue, even with existing externals.

But `{` and `}` are *not* FUDI-specific characters!

**Comment by reduzent** [comment]
> However, it is the _only_ way to do this without breaking existing patches.

What patches? Either they've been modified with external (to Pd) editors or they don't use any curly braces. Patches that receive curly braces from externals sources ([pdlua], network, serial, file, etc.) shouldn't break, assuming the curly braces would be escaped then.

> > As long as FUDI-specific special characters are escaped during FUDI-encoding, there should be no issue, even with existing externals.
> 
> But `{` and `}` are _not_ FUDI-specific characters!

Fair point. Considering the Pd "language" they've been non-existent and thus I think it's worthwhile thinking about introducing them now. 

Maybe I missed it before, but can you give an exact example of what would break?

**Comment by Spacechild1** [comment]
> What patches? Either they've been modified with external (to Pd) editors or they don't use any curly braces.

Pd messages do not have to be typed, they can be generated with list operations, read from a file, received from a network, etc. 

Here's a simple example that would break with @umlaeute's proposal:
[json-test.zip](https://github.com/pure-data/pure-data/files/10372291/json-test.zip)

> Considering the Pd "language" they've been non-existent and thus I think it's worthwhile thinking about introducing them now.

That's not true, either. Again, curly braces have never been a special character in Pd messages. The only reason why you can not type curly braces is that Pd did not have a proper escape mechanism for sending Pd messages between the core and the GUI. If such a mechanism had been in place from the beginning, there would have never been a limitation on the use of curly braces. So it is not a Pd *language* limitation at all, only a technical hindrance in the GUI implementation.

**Comment by reduzent** [comment]
> Pd messages do not have to be typed manually, they can be generated with list operations, read from a file, received from a network, etc.

Exactly my point. Curly braces from external sources would be treated as literal curly braces by escaping them (as it is the case for commas, semicolons, whitespaces). `[125(` -> `[list tosymbol]` would create a `symbol \}`, accordingly.

**Comment by Spacechild1** [comment]
> the curlies in the patch are always symbols, never special characters; so they should be fine.

Ah, right. Now imagine I replace `[l2s] -> [list fromsymbol]` with `[fudiparse]`; or I send the list of bytes over the network and the other end does `[fudiparse]`; or I write the Pd list to a file and someone else reads it back in.

If I understand correctly, what you imagine is that `binbuf_gettext()` would start to escape curly braces in symbols and `binbuf_text()` would interpret "isolated" curly braces as something like `A_CURLY`. This only works as long as all parties involved agree on the new protocol.

**Comment by reduzent** [comment]
> This only works as long as all parties involved agree on the new protocol.

Which parties? I'm not questioning, rather curious.

**Comment by Spacechild1** [comment]
Let's take my example patch and assume that I write the resulting Pd list `{ "foo" = 123 \, "bar" = 5 }` to a file. Now, 1 year in the future the meaning of (isolated) curly braces has changed. If I try to load this file with `[text]`, I do not get the same result. The curly braces in the message are instead parsed as special atoms (e.g. `A_CURLY`) and not as symbols.

Or see the `[fudiformat]` or `[fudiparse]` example above.

Generally, you cannot take an existing protocol, change the meaning of certain characters (outside escape sequences) and expect no breakage. That's just not possible.

**Comment by reduzent** [comment]
> Generally, you cannot take an existing protocol, change the meaning of certain characters (outside escape sequences) and expect no breakage. That's just not possible.

Valid point. Maybe this could be addressed with a compat flag? /me ducks from IOhannes...

I find the idea of multidimensional lists very intriguing and wouldn't want to bury it too lightly.

**Comment by Spacechild1** [comment]
> I find the idea of multidimensional lists very intriguing

Me too!

> wouldn't want to bury it too lightly.

Another possibility would be to use the dollar sign. There have been discussions about introducing the `$@` notation as a shorthand for "all (creation) arguments". Similary, we could introduce `${` and `$}` to denote the start/end of a nested list.

Now, *technically* - and unfortunately - you do not have to escape a dollar sign if the following character is not a number. `foo$bar` is the same as `foo\$bar`. The former is certainly bad practice and I'm not sure if this is formally documented, or just an implementation detail. 

What we *could* do is explicitly disallow the use of unescaped literal dollar signs to open up the possibility for future extensions of the Pd syntax. Of course, this might be disabled with a compatibility flag :-)

**Comment by porres** [comment]
> Another possibility would be to use the dollar sign. There have been discussions about introducing $@ for "all (creation) arguments". Similary, we could introduce ${ and $} to denote the start/end of a nested list.


Yeah, the tricks to use curly braces has already been used by power users and Pd can receive and deal with them when receiving messages over the network. I can confirm that sending /{foo,bar} from SuperCollider to Pd via OSC works fine and routeOSC can deal with it to match foo and bar.

It's weird that "{" on its own would have a special syntax meaning but not as part of symbols. It makes sense to me to take advantage of this existing special character and use it to expand it for possible syntax expansion.

**Comment by Spacechild1** [comment]
Oh, and you *can* already type curly braces in text windows. I totally forgot about this.

**Comment by porres** [comment]
> Oh, and you can already type curly braces in text windows. I totally forgot about this.

Yeah, we missed this here, but I mentioned it in the recent discussion on the pd-list

**Comment by umlaeute** [comment]
how about `[list flatten]` and `[list unflatten]`, that convert between nested lists and flat lists (with `{` and `}` being ordinary symbols)?

**Comment by umlaeute** [comment]
i was going to say, that `[text]` allows you to enter `$1` and get a literal symbol "$1".

but it seems it is much more complicated and convoluted: you *do* get an `A_DOLLAR` (resp `A_DOLLSYM`) and *sometimes* warnings.

i wonder how you could use such an `A_DOLLAR` retrieved from `[text]`.

**Comment by umlaeute** [comment]
conversely, you *can**not*** receive `$1` via `[netreceive]`. that looks broken to me as well.

if nobody can come up a use case for `[text]` outputting `A_DOLLAR`, probably `[netreceive]` and `[text]` should just output literals containing `$` (for my curly-suggestion that would also mean, that they would output literals containing `{}`)

**Comment by Spacechild1** [comment]
> i wonder how you could use such an A_DOLLAR retrieved from [text].

Dollars/dollarsyms in `[text]` have special meaning, they can be set dynamically with the `[text sequence]` object:

![grafik](https://user-images.githubusercontent.com/16126632/227971007-5818b253-0414-4206-8404-746880773604.png)

> you do get an A_DOLLAR (resp A_DOLLSYM) and sometimes warnings.

That's a bug! If lines with dollars/dollarsyms are obtained with `[text get]`, the dollars/dollarsyms should be bashed to symbols.

**Comment by umlaeute** [comment]
ah yes. thx.

**Comment by Spacechild1** [comment]
> conversely, you cannot receive $1 via [netreceive]. that looks broken to me as well.

What is broken about it? Currently you get a warning (`netreceive: got dollar sign in message`). The only other option would be to bash dollars/dollarsym to symbols. Both seem equally legitimate to me. What else could `[netreceive]` do with actual dollars/dollarsyms? Replace them with the current canvas arguments? Maybe, but I don't really see a potential use case for this.

**Comment by porres** [comment]
Hi, I am revisiting this thread after 3 years :) because of the dict support discussion and all.

I think we got carried away with parallel discussions.

Some considerations from this conversation.

"{" and "}" are already supported in [text], they are manageable in Vanilla with obscure tricks using [makefilename %c]. They are also fully supported to be typed in PlugData and Pd-L2ork, and generate Vanilla compatible patches that can be opened. See:

<img width="387" height="221" alt="Image" src="https://github.com/user-attachments/assets/efd36279-3d5c-4ada-ad19-6f91175cb0b4" />

by the way, the patch is this

```
#N canvas 827 239 734 565 12;
#X msg 134 93 I created { this message } in { PlugData };
```

In https://github.com/pure-data/pure-data/issues/505#issuecomment-1374880190 

we have the following remark as a possible hurdle

"escaping of curlies when (re)storing files (they shouldn't be escaped at all); this might already be solved (i haven't checked)"

So, well, it seems it is solved. I could open the patch above, edit the message, and open back in Vanilla and PlugData just fine...

<img width="435" height="291" alt="Image" src="https://github.com/user-attachments/assets/5faae0ae-c204-4a65-aaa2-296cf3f61f15" />

Which narrows down to simply 

- [ ] drop the filter and allow input of curlies (gui -> communication) (which is simple to solve)

-----------------------------

We have discussed using this character as a new syntactic goodie, but it could break things in edge cases, requiring compatibility flags and stuff.

Instead, some objects could simply work with this symbol as a special character and perform miracles, such as proposed in 

https://github.com/pure-data/pure-data/issues/505#issuecomment-1400748312

This is also what I'd like to do in a dict external or something.

What I took from this discussion is that we couldn't agree on an easy and satisfactory way to use this character into a new syntax, so I guess we can just let it go.

To me, expanding Pd's syntax could be treated as a separate issue/discussion/request that has nothing to do with this one, which is simply a request to finally allow typing { } with Pd's GUI.

And it's not like it's the end of the world for expanding Pd's syntax. See https://github.com/pure-data/pure-data/issues/505#issuecomment-1375625695 as an alternative that could be implemented and discussed separately.

What do you say?

I could try a PR that simply removes the filter to close this issue.
